plugins/qcheck-stm/src/ir_of_gospel.ml

```diff
@@ -76,11 +76,35 @@ let ty_var_substitution config (vd : val_description) =
   let value_name, value_type = (vd.vd_name.id_str, vd.vd_type) in
   assert (is_a_function value_type);
   let open Ppxlib in
+  let open Reserr in
   let ret = function None -> Reserr.ok [] | Some x -> Reserr.ok x in
+  let rec check pos ty =
+    match ty.ptyp_desc with
+    | Ptyp_constr (_, args) -> (
+        match List.find_opt (Cfg.is_sut config) args with
+        | Some _ -> (
+            match pos with
+            | `Left -> error (Sut_as_type_inst value_name, ty.ptyp_loc)
+            | `Right -> error (Returning_sut value_name, ty.ptyp_loc))
+        | None ->
+            let* _ = List.map (check pos) args |> sequence in
+            ok ())
+    | Ptyp_tuple elems -> (
+        match List.find_opt (Cfg.is_sut config) elems with
+        | Some _ -> (
+            match pos with
+            | `Left -> error (Sut_in_tuple value_name, ty.ptyp_loc)
+            | `Right -> error (Returning_sut value_name, ty.ptyp_loc))
+        | None ->
+            let* _ = List.map (check pos) elems |> sequence in
+            ok ())
+    | _ -> ok ()
+  in
   let rec aux seen ty =
     match ty.ptyp_desc with
     | Ptyp_any | Ptyp_var _ -> ret seen
     | Ptyp_arrow (_, l, r) ->
+        let* _ = check `Left l in
```

Comment:
The expression evaluates to a `unit reserr`, so as well have `let* () =` rather than a wild pattern.